src/storage/heap_file.c

```diff
@@ -24743,5 +24748,287 @@ heap_nonheader_page_capacity ()
+
+  // Add new links to the last page of the heap.
+  error_code = heap_add_chain_links (thread_p, hfid, &heap_last_page_vpid, &heap_pages_array[0], NULL,
+                                     &heap_last_page_watcher, true, true);
+  if (error_code != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      goto cleanup;
+    }
+
+  // Now update the last page of the heap header.
+
+  heap_hdr->estimates.last_vpid = heap_pages_array[array_size - 1];
+  heap_hdr->estimates.num_pages += (int) array_size;
+
+  // Log this change.
+  addr.pgptr = heap_header_watcher.pgptr;
+  addr.vfid = &hfid->vfid;
+  addr.offset = HEAP_HEADER_AND_CHAIN_SLOTID;
+
+  log_append_undoredo_data (thread_p, RVHF_STATS, &addr, sizeof(HEAP_HDR_STATS), sizeof (HEAP_HDR_STATS), 
+                            &heap_hdr_prev, heap_hdr);
+
+  // Set the page as dirty.
+  pgbuf_set_dirty (thread_p, heap_header_watcher.pgptr, DONT_FREE);  
+
+cleanup:
+
+   if (page_watcher.pgptr)
+    {
+      pgbuf_ordered_unfix_and_init (thread_p, page_watcher.pgptr, &page_watcher);
+    }
+
+  if (heap_last_page_watcher.pgptr)
+    {
+      pgbuf_ordered_unfix_and_init (thread_p, heap_last_page_watcher.pgptr, &heap_last_page_watcher);
+    }
+
+  if (heap_header_watcher.pgptr)
+    {
+      pgbuf_ordered_unfix_and_init (thread_p, heap_header_watcher.pgptr, &heap_header_watcher);
+    }
+
+  // Check if we have errors to abort the sysop.
+  if (error_code != NO_ERROR)
+    {
+      // Safeguard
+      ASSERT_ERROR ();
+      log_sysop_abort (thread_p);
+    }
+  else
+    {
+      // Commit the sysop
+      log_sysop_commit (thread_p);
+    }
+
+  return error_code;
+}
+
+static int
+heap_get_page_with_watcher (THREAD_ENTRY * thread_p, const VPID *page_vpid, PGBUF_WATCHER * pg_watcher)
+{
+  int error_code = NO_ERROR;
+
+   // Safeguards.
+  assert (pg_watcher != NULL);
+  assert (page_vpid != NULL);
+
+  pg_watcher->pgptr = heap_scan_pb_lock_and_fetch (thread_p, page_vpid, OLD_PAGE, X_LOCK, NULL, pg_watcher);
+  if (pg_watcher->pgptr == NULL)
+    {
+      ASSERT_ERROR_AND_SET (error_code);
+      return error_code;
+    }
+
+   return error_code;
+}
+
+static int
+heap_add_chain_links (THREAD_ENTRY * thread_p, const HFID * hfid, const VPID * vpid, const VPID * next_link,
+                      const VPID * prev_link, PGBUF_WATCHER * page_watcher, bool keep_page_fixed,
+                      bool is_page_watcher_inited)
+{
+  LOG_DATA_ADDR addr = LOG_DATA_ADDR_INITIALIZER;
+  int error_code = NO_ERROR;
+
+  // Init watcher if needed.
+  if (!is_page_watcher_inited)
+    {
+      PGBUF_INIT_WATCHER (page_watcher, PGBUF_ORDERED_HEAP_NORMAL, hfid);
+
+       // Get a watcher for this page.
+      error_code = heap_get_page_with_watcher (thread_p, vpid, page_watcher);
+      if (error_code != NO_ERROR)
+        {
+          ASSERT_ERROR ();          
+          return error_code;
+        }
+    }
+
+  // Make sure we fixed the page.
+  assert (pgbuf_is_page_fixed_by_thread (thread_p, (VPID *) vpid));
```

Comment:
Cast won't be necessary if pgbuf_is_page_fixed_by_thread signature is changed to accept `const VPID *`.

Reply:
```suggestion
  assert (pgbuf_is_page_fixed_by_thread (thread_p, vpid));
```